csrc/multidevice/utils.cpp

```diff
@@ -142,19 +142,49 @@ int64_t numDeviceDims(TensorView* tv) {
 }
 
 bool isResharding(Expr* expr) {
-  std::unordered_set<TensorView*> tvs;
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
-    tvs.insert(tv);
-  }
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
-    tvs.insert(tv);
-  }
-  if (tvs.empty()) {
-    return false;
+  // we don't use getTvsWithDifferentSharding because it creates a computeAtMap,
+  // which is too costly
+  for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      // exit early in the unsharded case for performance
+      if (!producer->hasDeviceMesh() && !consumer->hasDeviceMesh()) {
+        continue;
+      }
+      // If device mesh are different, the Expr is resharding
+      if (producer->getDeviceMesh().vector() !=
+          consumer->getDeviceMesh().vector()) {
+        return true;
+      }
+
+      // Create a map between producer's and consumer's IterDomains. We iterate
+      // over producer's iterdomain and compare sharding type with consumer's
+      // iterdomain
+      const auto p2c_map =
+          PairwiseRootDomainMap(producer, consumer).mapProducerToConsumer();
+      for (auto p_id :
+           TensorDomain::noReductions(producer->getMaybeRFactorDomain())) {
+        auto p2c_map_it = p2c_map.find(p_id);
+        if (p2c_map_it == p2c_map.end()) {
+          NVF_ERROR(
+              !p_id->isDeviceDim(),
+              "the producer ",
+              producer,
+              " has a sharded dimension ",
+              p_id,
+              " that is not mapped to its consumer ",
+              consumer);
+          continue;
+        }
+        auto c_id = p2c_map_it->second;
+        if (p_id->getParallelType() != c_id->getParallelType() &&
+            (p_id->isDeviceDim() || c_id->isDeviceDim())) {
+          // Mismatch found
+          return true;
+        }
+      }
+    }
   }
-  auto tv_ref = *tvs.begin();
-  tvs.erase(tv_ref);
-  return !getTvsWithDifferentSharding(tv_ref, tvs).empty();
```

Comment:
getTvsWithDifferentSharding has several other uses than isResharding. What about fixing getTvsWithDifferentSharding instead?

Reply:
`getTvsWithDifferentSharding` inputs tv are meant to be very general, with no a priori assumption on producer/consumer dependency -- on the contrary, in `isResharding` we have the extra info that the tvs we are considering are I/O of the expr.

While it's good to have a general helper function like `getTvsWithDifferentSharding`, it also makes sense to have an optimized implementation for special cases like the one of `isResharding`

So, I agree with you, fixing `getTvsWithDifferentSharding` would be more general, but I am not sure it is possible. Let me know if you think there is a way

Reply:
Interesting. All the non-test uses are in https://github.com/NVIDIA/Fuser/blob/b020415858caa38a77b3c309566903f3fbdbac3d/csrc/multidevice/utils.cpp and they all reason about producer/consumer. While there probably is a value of keeping a general helper like `getTvsWithDifferentSharding` (at least the general form seems to be useful for https://github.com/NVIDIA/Fuser/blob/b020415858caa38a77b3c309566903f3fbdbac3d/tests/cpp/test_resharding.cpp#L284), other uses in multidevice/utils.cpp are still worth fixing. One possible way is to create a helper like `isResharding(input_tv, output_tv)` and call that instead.

Reply:
We could check if `input_tv` and `output_tv` are producer consumer, and use the quick `PairwiseRootDomainMap` path if true, otherwise use the expensive `ComputeAt` map?

Reply:
I introduced a new function `bool haveDifferentShardings(TensorView* producer, TensorView* consumer)` in `multidevice/utils.(h|cpp)` that I use instead of calling `getTvsWithDifferentSharding` everywhere in the source code, so we should be good.

I only let the test in `test_resharding.cpp` calling `getTvsWithDifferentSharding`

> We could check if input_tv and output_tv are producer consumer, and use the quick PairwiseRootDomainMap path if true, otherwise use the expensive ComputeAt map?

I did not implement that for now, we can do it later if we feel the need.

Wdyt ?